pkg/apiserver/registry/generic/key.go

```diff
@@ -0,0 +1,128 @@
+package generic
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/validation/path"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+type Key struct {
+	Group     string
+	Resource  string
+	Namespace string
+	Name      string
+}
+
+// KeyRootFunc is used by the generic registry store to construct the first portion of the storage key.
+func KeyRootFunc(ctx context.Context, gr schema.GroupResource) string {
+	ns, _ := genericapirequest.NamespaceFrom(ctx)
+	key := &Key{
+		Group:     gr.Group,
+		Resource:  gr.Resource,
+		Namespace: ns,
+	}
+	return key.String()
+}
+
+// NamespaceKeyFunc is the default function for constructing storage paths to
+// a resource relative to the given prefix enforcing namespace rules. If the
+// context does not contain a namespace, it errors.
+func NamespaceKeyFunc(ctx context.Context, gr schema.GroupResource, name string) (string, error) {
+	ns, ok := genericapirequest.NamespaceFrom(ctx)
+	if !ok || len(ns) == 0 {
+		return "", apierrors.NewBadRequest("Namespace parameter required.")
+	}
+	if len(name) == 0 {
+		return "", apierrors.NewBadRequest("Name parameter required.")
+	}
+	if msgs := path.IsValidPathSegmentName(name); len(msgs) != 0 {
+		return "", apierrors.NewBadRequest(fmt.Sprintf("Name parameter invalid: %q: %s", name, strings.Join(msgs, ";")))
+	}
+	key := &Key{
+		Group:     gr.Group,
+		Resource:  gr.Resource,
+		Namespace: ns,
+		Name:      name,
+	}
+	return key.String(), nil
+}
+
+// NoNamespaceKeyFunc is the default function for constructing storage paths
+// to a resource relative to the given prefix without a namespace.
+func NoNamespaceKeyFunc(ctx context.Context, prefix string, gr schema.GroupResource, name string) (string, error) {
+	if len(name) == 0 {
+		return "", apierrors.NewBadRequest("Name parameter required.")
+	}
+	if msgs := path.IsValidPathSegmentName(name); len(msgs) != 0 {
+		return "", apierrors.NewBadRequest(fmt.Sprintf("Name parameter invalid: %q: %s", name, strings.Join(msgs, ";")))
+	}
+	key := &Key{
+		Group:    gr.Group,
+		Resource: gr.Resource,
+		Name:     name,
+	}
+	return prefix + key.String(), nil
+}
+
+func ParseKey(key string) (*Key, error) {
```

Comment:
there currently isn't existing test coverage for parsing and generating strings. i'd like to add coverage in a follow-up PR, but skipped i skipped it for now so this PR can merge and unblock @bergquist.